### DISCUSSION TITLE
How to get histogram of colors present in an image

### DISCUSSION THREAD

**Discussion by ORESoftware** [ImageMagick/ImageMagick#3033] [Consulting]
I am creating an image app, where users can find an image by dominant color. I am not sure how to go about this yet, but here is my thinking:

1. when they upload a new image, my app creates a histogram of 32 or 64 colors, forcing each pixel into the buckets.
2. when the user wants to look up images by dominant color, they select one of the 32 or 64 colors, and we look up images that match that, sorted by highest match rate, etc.

any idea how to implement? Seems like the bucket thing is necessary, but for each pixel, may need to put into more than one bucket, etc. not sure how colors work in images.

**Comment by snibgo** [comment]
I suppose you have a standard list of N colours, where N=32 or N=64. This would be an image of N pixels.

You can "-remap" an image to those N colours. This change each pixel to the closest of those N colours. Then "-verbose info:" will count how many pixels are of each colour. If images can be different sizes, then you might convert each count to a percentage of the size in pixels.

> ... but for each pixel, may need to put into more than one bucket ...

Why? I don't understand that. Perhaps you are thinking that a pixel might fall between two standard colours. True, it might. IM can handle that: "-remap" with dithering (which is the default). Then if many identical pixels are exactly half way between colours A and B, then half will be remapped to A, and half to B.

**Comment by ORESoftware** [comment]
Yeah, if we have 7 buckets (ROYGBIV), then a pixel could be 70% R, 20% O, and 35% Y, or something. (The percentages don't have to add up there).  Just trying to figure out how to store the data in a format that can be searched.

**Comment by snibgo** [comment]
Ah, that would be different. Okay, then you need a colorspace with as many channels as you have standard colours. In computers, additive colours usually have only three channels. (Subtractive colours have four channels).

Sadly, IM can't readily handle 7 channels (or 32 or 64). G'MIC can do that.

You would need to decide how you calculate those percentages, 70% of Red, 20% of Orange etc. That's not a maor problem.

But then you would need to decide how to reduce that huge amount of data to a searchable scheme.

**Comment by fmw42** [comment]
All your images would need to have the set of 32 or 64 colors. In imagemagick you can create a color table image of your desired 64 colors. Then map that onto every image. See +dither -remap at https://legacy.imagemagick.org/Usage/quantize/#remap

Now you can histogram your image so that they all are compared to the same set of colors.
